src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StatusAndTiming.java

```diff
@@ -0,0 +1,314 @@
+ * </ol>
+ * <p/> <strong>Parallel branch handling:</strong>
+ * <ol>
+ *     <ul>Each branch is considered independent</ul>
+ *     <ul>Branches may succeed, fail, or be in-progress/waiting for input.</ul>
+ * </ol>
+ * @author Sam Van Oort
+ */
+public class StatusAndTiming {
+
+    /**
+     * Check that all the flownodes & run describe the same pipeline run/execution
+     * @param run Run that nodes must belong to
+     * @param nodes Nodes to match to run
+     * @throws IllegalArgumentException For the first flownode that doesn't belong to the FlowExectuon of run
+     */
+    public static void verifySameRun(@Nonnull WorkflowRun run, @CheckForNull FlowNode... nodes) throws IllegalArgumentException {
+        if (nodes == null || nodes.length == 0) {
+            return;
+        }
+        FlowExecution exec = run.getExecution();
+        int i =0;
+        for (FlowNode n : nodes) {
+            if (n!= null && n.getExecution() != exec) {
+                throw new IllegalArgumentException("FlowNode not part of the same execution found, at index "+i+" with ID "+n.getId());
+            }
+            i++;
+        }
+    }
+
+    /** Return true if the run is paused on input */
+    public static boolean isPendingInput(WorkflowRun run) {
+        // Logic borrowed from Pipeline Stage View plugin, RuneEx
+        InputAction inputAction = run.getAction(InputAction.class);
+        if (inputAction != null) {
+            List<InputStepExecution> executions = inputAction.getExecutions();
+            if (executions != null && !executions.isEmpty()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Nonnull
+    public static GenericStatus computeChunkStatus(@Nonnull WorkflowRun run,
+                                                   @CheckForNull FlowNode before, FlowChunk chunk, @CheckForNull FlowNode after) {
+        FlowExecution exec = run.getExecution();
+        if (exec == null) {
+            return GenericStatus.NOT_EXECUTED;
+        }
+        return computeChunkStatus(run, before, chunk.getFirstNode(exec), chunk.getLastNode(exec), after);
+    }
+
+    /**
+     * Compute the overall status for a chunk comprising firstNode through lastNode, inclusive
+     * <p/> All nodes must be in the same execution
+     * <p/> Note: for in-progress builds with parallel branches, if the branch is done, it has its own status.
+     * @param run Run that nodes belong to
+     * @param before Node before the first node in this piece
+     * @param firstNode First node of this piece
+     * @param lastNode Last node of this piece (if lastNode==firstNode, it's a single FlowNode)
+     * @param after Node after this piece, null if the lastNode is the currentHead of the flow
+     * @return Status for the piece, or {@link GenericStatus#UNKNOWN} if the FlowExecution is null.
+     */
+    @Nonnull
+    public static GenericStatus computeChunkStatus(@Nonnull WorkflowRun run,
```

Comment:
@vivek Valid -- I think I'm going to address this a the follow-on PR which more closely couples with the chunked visitor.  I'd like to be able to tie together the changes there.